File: code/js/src/Domain/wrappers/email/EmailValidator.ts
import { EmailValidationError } from './EmailValidationError';

export class EmailValidator {
    private static readonly MAX_EMAIL_LENGTH = 50;
    private static readonly MIN_EMAIL_LENGTH = 8;
    private static readonly EMAIL_REGEX =
        /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}$/;

    public constructor(
        readonly maxEmailLength: number = EmailValidator.MAX_EMAIL_LENGTH,
        readonly minEmailLength: number = EmailValidator.MIN_EMAIL_LENGTH,
        readonly emailRegex: RegExp = EmailValidator.EMAIL_REGEX,
    ) {}

    validate(email: string): EmailValidationError[] {
        const errors: EmailValidationError[] = [];

        if (email.trim() === '') {
            errors.push(EmailValidationError.Blank);
        }

        if (!email.match(this.emailRegex)) {
            errors.push(EmailValidationError.InvalidFormat);
        }

        if (
            email.length < this.minEmailLength ||
            email.length > this.maxEmailLength
        ) {
            errors.push(
                new EmailValidationError.InvalidLength(
                    this.minEmailLength,
                    this.maxEmailLength,
                ),
            );
        }

        if (errors.length > 0) {
            return errors;
        }

        return [];
    }
}

File: code/server/jvm/im-domain/src/main/kotlin/im/domain/wrappers/email/EmailValidator.kt
package im.domain.wrappers.email

import im.domain.Either
import im.domain.failure
import im.domain.success

class EmailValidator(
    private val maxEmailLength: Int = MAX_EMAIL_LENGTH,
    private val minEmailLength: Int = MIN_EMAIL_LENGTH,
    private val emailRegex: String = EMAIL_REGEX,
) {
    companion object {
        private const val MAX_EMAIL_LENGTH = 50
        private const val MIN_EMAIL_LENGTH = 8
        private const val EMAIL_REGEX = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Z|a-z]{2,}\$"
    }

    fun validate(email: String): Either<List<EmailValidationError>, Unit> {
        val errors = mutableListOf<EmailValidationError>()

        if (email.isBlank()) {
            errors.add(EmailValidationError.Blank)
        }

        if (!email.matches(Regex(emailRegex))) {
            errors.add(EmailValidationError.InvalidFormat)
        }

        if (email.length !in minEmailLength..maxEmailLength) {
            errors.add(EmailValidationError.InvalidLength(minEmailLength, maxEmailLength))
        }

        if (errors.isNotEmpty()) {
            return failure(errors)
        }

        return success(Unit)
    }
}
